source/extensions/upstreams/http/tcp/upstream_request.h

```diff
@@ -26,6 +26,7 @@ class TcpConnPool : public Router::GenericConnPool, public Envoy::Tcp::Connectio
               Upstream::ResourcePriority priority, Upstream::LoadBalancerContext* ctx) {
     conn_pool_data_ = thread_local_cluster.tcpConnPool(priority, ctx);
   }
+  ~TcpConnPool() override { cancelAnyPendingStream(); }
```

Comment:
as said in the issue can we ENVOY_BUG if there's streams remaining?
/wait

Reply:
Sorry this was unclear, I meant if there was an actual reference that was getting cleaned up - if upstream_handle_ is non-null.

AFIK the pool should be getting onPoolFailure before getting torn down.  If you have a repeatable case where that's not happening not using in-house code I'd be interested in understanding the lifetime issue.  Otherwise if we're going to avoid the UAF we should indicate it's a bug that we're not being shut down as expected.